PowerUp/app/src/main/java/powerup/systers/com/MapActivity.java

```diff
@@ -26,30 +29,37 @@
 
 public class MapActivity extends Activity {
 
-    private DatabaseHandler mDbHandler;
+    private DataSource dataSource;
     private OnClickListener onClickListener = new OnClickListener() {
         @Override
         public void onClick(View v) {
-            ImageView scenarioChooser = (ImageView) v;
+            final ImageView scenarioChooser = (ImageView) v;
             if (v.isEnabled()){
                 // checks whether has been already completed & then open GameActivity if setSessionId returns true
-                if (getmDbHandler().setSessionId(getScenarioName(scenarioChooser.getId()))) {
-                    startActivityForResult(new Intent(MapActivity.this, GameActivity.class), 0);
-                    overridePendingTransition(R.animator.fade_in_custom, R.animator.fade_out_custom);
-                } else if (new MinesweeperSessionManager(MapActivity.this).isMinesweeperOpened()) { //if minesweeper game was left incomplete
-                    startActivity(new Intent(MapActivity.this, MinesweeperGameActivity.class));
-                    overridePendingTransition(R.animator.fade_in_custom, R.animator.fade_out_custom);
-                } else if (new SinkToSwimSessionManager(MapActivity.this).isSinkToSwimOpened()) {
-                    startActivity(new Intent(MapActivity.this, SinkToSwimGame.class));
-                } else if (new VocabMatchSessionManager(MapActivity.this).isVocabMatchOpened()) {
-                    startActivity(new Intent(MapActivity.this, VocabMatchGameActivity.class));
-                } else {
-                    Intent intent = new Intent(MapActivity.this, ScenarioOverActivity.class);
-                    intent.putExtra(PowerUpUtils.SOURCE,PowerUpUtils.MAP);
-                    intent.putExtra(PowerUpUtils.SCENARIO_NAME,getScenarioName(scenarioChooser.getId()));
-                    startActivityForResult(intent, 0);
-                    overridePendingTransition(R.animator.fade_in_custom, R.animator.fade_out_custom);
-                }
+                dataSource.setSessionId(getScenarioName(scenarioChooser.getId()), new IDataSource.LoadBooleanCallback() {
+                    @Override
+                    public void onResultLoaded(boolean value) {
+                        if(value) {
+                            startActivityForResult(new Intent(MapActivity.this, GameActivity.class), 0);
+                            overridePendingTransition(R.animator.fade_in_custom, R.animator.fade_out_custom);
+                        }
+                        else if (new MinesweeperSessionManager(MapActivity.this).isMinesweeperOpened()) { //if minesweeper game was left incomplete
```

Comment:
Here, I would suggest, to modularize code little bit, will be easy for you to write Unit test for this. 
So looks like, else ifs are not dependent on value, so I would suggest put that inside 
if(value) {
// no change
} else
 call private method()

// implement private method 
 // checking if Minesweeper opened/SiktoSwimOpened/isVocalMatchOpened